src/map/achievement.c

```diff
@@ -0,0 +1,1399 @@
+		else
+			achievement->progress_set(sd, ad, index, progress);
+	}
+
+	return true;
+}
+
+/**
+ * Validates monster kill type objectives.
+ * @type ACH_KILL_MOB_CLASS
+ * @param sd       pointer to session data.
+ * @param mob_id   (criteria) class of the monster checked for.
+ * @param progress (goal) progress to be added.
+ * @see achievement_vaildate_type()
+ */
+void achievement_validate_mob_kill(struct map_session_data *sd, int mob_id)
+{
+	struct achievement_objective criteria = { 0 };
+
+	nullpo_retv(sd);
+
+	Assert_retv(mob_id > 0 && mob->db(mob_id) != NULL);
+
+	criteria.mobid = mob_id;
+
+	achievement->validate_type(sd, ACH_KILL_MOB_CLASS, &criteria, 1, true);
+}
+
+/**
+ * Validate monster damage type objectives.
+ * @types ACH_DAMAGE_MOB_REC_MAX
+ *        ACH_DAMAGE_MOB_REC_TOTAL
+ * @param sd       pointer to session data.
+ * @param damage   amount of damage received/dealt.
+ * @param received received/dealt boolean switch.
+ */
+void achievement_validate_mob_damage(struct map_session_data *sd, unsigned int damage, bool received)
+{
+	struct achievement_objective criteria = { 0 };
+
+	nullpo_retv(sd);
+	Assert_retv(damage > 0);
+
+	if (received) {
+		achievement->validate_type(sd, ACH_DAMAGE_MOB_REC_MAX, &criteria, damage, false);
+		achievement->validate_type(sd, ACH_DAMAGE_MOB_REC_TOTAL, &criteria, damage, true);
+	} else {
+		achievement->validate_type(sd, ACH_DAMAGE_MOB_MAX, &criteria, damage, false);
+		achievement->validate_type(sd, ACH_DAMAGE_MOB_TOTAL, &criteria, damage, true);
+	}
+
+}
+
+/**
+ * Validate player kill (PVP) achievements.
+ * @types ACH_KILL_PC_TOTAL
+ *        ACH_KILL_PC_JOB
+ *        ACH_KILL_PC_JOBTYPE
+ * @param sd       pointer to killed player's session data.
+ * @param dstsd    pointer to killer's session data.
+ */
+void achievement_validate_pc_kill(struct map_session_data *sd, struct map_session_data *dstsd)
+{
+	struct achievement_objective criteria = { 0 };
+
+	nullpo_retv(sd);
+	nullpo_retv(dstsd);
+
+	/* */
+	achievement->validate_type(sd, ACH_KILL_PC_TOTAL, &criteria, 1, true);
+
+	/* */
+	VECTOR_INIT(criteria.jobid);
+	VECTOR_ENSURE(criteria.jobid, 1, 1);
+	VECTOR_PUSH(criteria.jobid, dstsd->status.class);
+
+	/* Job class */
+	achievement->validate_type(sd, ACH_KILL_PC_JOB, &criteria, 1, true);
+	/* Job Type */
+	achievement->validate_type(sd, ACH_KILL_PC_JOBTYPE, &criteria, 1, true);
+
+	VECTOR_CLEAR(criteria.jobid);
+}
+
+/**
+ * Validate player kill (PVP) achievements.
+ * @types ACH_DAMAGE_PC_MAX
+ *        ACH_DAMAGE_PC_TOTAL
+ *        ACH_DAMAGE_PC_REC_MAX
+ *        ACH_DAMAGE_PC_REC_TOTAL
+ * @param sd       pointer to source player's session data.
+ * @param dstsd    pointer to target player's session data.
+ * @param damage   amount of damage dealt / received.
+ */
+void achievement_validate_pc_damage(struct map_session_data *sd, struct map_session_data *dstsd, unsigned int damage)
+{
+	struct achievement_objective criteria = { 0 };
+
+	nullpo_retv(sd);
+	nullpo_retv(dstsd);
+
```

Comment:
this one never crash inside func1 even if ptr is NULL
```
void func1(int *ptr)
{
  func2(ptr);
}
```

this one will crash if ptr is NULL
```
void func3(int *ptr)
{
  *ptr = 10;
}
```

This mean in func1 check ptr for NULL is useless. It only slow down execution.
But in func3 must be check for NULL, or it may crash.

I saw in many places where was useless nullpo checks in previous code.

Useless nullpo checks is not critical, but for performance better not add it.